### ISSUE TITLE
Lifecycle: Allow `locking` for setting/getting routers on inbounds

### ISSUE THREAD

**Issue by willhug** [yarpc/yarpc-go#913]
Noticed on the GRPC diff that we need a lock around setting an inbound's router, since the Lifecycle sync is used to managed start/stop semantics and having a lock around the router is another layer on top of that, we should probably isolate that usage.

**Comment by kriskowal** [comment]
Can we just clarify that SetRouter must be called in the main thread before starting anything?

**Comment by peter-edge** [comment]
This is somewhat a duplicate of https://github.com/yarpc/yarpc-go/issues/825 I believe @willhug @kriskowal

**Comment by willhug** [comment]
This should be all semantic, and once we're done #1189 it'll be less of an issue because we won't be calling "procedures" on the router.
